### ISSUE TITLE
set up Travis CI

### ISSUE THREAD

**Issue by derekbruening** [DynamoRIO/dynamorio#1962]
Our Chromium-infrastructure buildbots are a pain to maintain with all of the Chromium infrastructure changes.  The proposal is to switch to a combination of distributed developer-owned machines submitting to CDash as well as setting up Travis CI for continuous integration testing.  Travis supports Linux and most importantly Mac, to give us coverage there.  Windows, ARM, and AArch64 will have to rely solely on CDash.

For nightly public builds we have not yet decided on a final solution.

**Comment by derekbruening** [comment]
I'm documenting here the error I'm seeing with gcc (instead of clang) on the Travis OSX test machines:

https://travis-ci.org/DynamoRIO/dynamorio/jobs/142318932

```
-- The ASM_NASM compiler identification is unknown
-- Found assembler: /usr/local/bin/nasm
-- Found nasm: /usr/local/bin/nasm
CMake Error at CMakeLists.txt:579 (message):
  gcc is required to build
-- Configuring incomplete, errors occurred!
```

clang works (https://travis-ci.org/DynamoRIO/dynamorio/jobs/142318931), so
I don't quite understand:

```
  if (NOT CMAKE_COMPILER_IS_GNUCC)
    # we use gcc extensions
    message(FATAL_ERROR "gcc is required to build")
```

But the plan is to just disable gcc on OSX and only test clang there: not worth the time to investigate further.

**Comment by derekbruening** [comment]
Here is the link for build results (I will set email to go to the -devs list once the configs are working):

https://travis-ci.org/DynamoRIO/dynamorio/builds

**Comment by derekbruening** [comment]
Hmm, we have quite a few failures now that I have the tests running:

https://travis-ci.org/DynamoRIO/dynamorio/builds/142353103

```
arm-debug-internal-64: skipped (cross-compile configuration unavailable)
android-release-external-32: skipped (cross-compile configuration unavailable)
arm-debug-internal-32: skipped (cross-compile configuration unavailable)
arm-release-external-64: skipped (cross-compile configuration unavailable)
arm-release-external-32: skipped (cross-compile configuration unavailable)
android-debug-internal-32: skipped (cross-compile configuration unavailable)

debug-internal-32: 212 tests passed, **** 8 tests failed: ****
	code_api|client.modules 
	code_api|client.exception =>  Application /home/travis/build/DynamoRIO/dynamorio/build_debug-internal-32/suite/tests/bin/simple_app (12864).  Internal Error: DynamoRIO debug check failure: /home/travis/build/DynamoRIO/dynamorio/core/unix/loader_linux.c:378 ti-ti_module  tls_info.num_mods 
	code_api|client.nudge_test 
	code_api|client.thread 
	code_api|client.pcache-use 
	code_api|client.nudge_ex 
	code_api|client.destructor =>  hasdtr::hasdtrApplication /home/travis/build/DynamoRIO/dynamorio/build_debug-internal-32/suite/tests/bin/client.destructor (12980).  Internal Error: DynamoRIO debug check failure: /home/travis/build/DynamoRIO/dynamorio/core/unix/loader_linux.c:378 ti-ti_module  tls_info.num_mods 
	code_api|client.drx-test 

debug-internal-64: 209 tests passed, **** 4 tests failed, of which 1 were flaky: ****
	code_api|common.decode 
	code_api|client.abort 
	code_api|client.thread 
	code_api|pthreads.ptsig_FLAKY 

release-external-32: build successful; no tests for this build
release-external-64: build successful; no tests for this build
vmsafe-debug-internal-32: build successful; no tests for this build
vps-debug-internal-32: build successful; no tests for this build
```

**Comment by derekbruening** [comment]
OK that's building with clang.  Building with gcc we just have:

```
arm-debug-internal-64: skipped (cross-compile configuration unavailable)
android-release-external-32: skipped (cross-compile configuration unavailable)
arm-debug-internal-32: skipped (cross-compile configuration unavailable)
arm-release-external-64: skipped (cross-compile configuration unavailable)
arm-release-external-32: skipped (cross-compile configuration unavailable)
android-debug-internal-32: skipped (cross-compile configuration unavailable)
debug-internal-32: 240 tests passed, **** 1 tests failed: ****
	code_api|client.modules 
debug-internal-64: all 232 tests passed
release-external-32: build successful; no tests for this build
release-external-64: build successful; no tests for this build
vmsafe-debug-internal-32: build successful; no tests for this build
vps-debug-internal-32: build successful; no tests for this build
```

It's due to there being no 32-bit libz.  Maybe the thing to do is file a separate issue on the clang failures, turn the clang travis suite into a build-only suite (may require some hacky runsuite params...), and fix client.modules.

**Comment by derekbruening** [comment]
#1801 covers clang failures

**Comment by derekbruening** [comment]
I closed this as the basics are in:

+ Run test suite on Mac
+ Run test suite w/ gcc on Linux
+ Run test suite (build-only) w/ clang on Linux
+ Notify the devs list on failures or changes in status

I will add some info/pointers to the relevant wiki pages.

Beyond that, we can file separate issues for significant additions or changes.  These might include:

+ Add A32 + A64 cross-compile builds to Travis (i.e., install the cross-compilers in the install rule, but we should consider the total build time: maybe just put in the clang build which is in parallel w/ the much longer gcc run?)

+ Use pull requests as trybots (too bad there's no Windows -- but could be useful for Mac)

**Comment by derekbruening** [comment]
Another piece the Chromium bots gave us that Travis might be able to provide if we are happy with any of its storage solutions:

+ Add nightly package builds

**Comment by derekbruening** [comment]
I updated https://github.com/DynamoRIO/dynamorio/wiki/Test-Suite
